### PR TITLE
Beta push for 1.0.16 (many fix and enhancement see details)

### PR DESCRIPTION
**Pull request by hthetiot**
It's time to deal with push requests and opened issues.

I will try to fix as mush bug and do small enhancements under review of @omochi and authority of @EricDavies.

- Add getSourceList and getAudioSourceList #80
- easyrtc screen capture chrome extension example. #87
- JSHint on easyrtc_int.js #169 
- Firefox and previous WebRTC impl does not call emitOnStreamClosed #173 
- getUserMediaConstraints() bug #176 
- Possible bad userAgent RegExp in Adapter.js #177
- A mistake in documentation #188
- wrong condition expression in buildLocalMediaStream #189
- missing argument for wasAcceptedCB #190
- too many arguments passed into onRemoveStreamHelper #192 
- using undefined error code SIGNAL_ERROR #193
- using undefined error code NOVIABLEICE #193
- Possible undefined getUserMedia #194
- Adapter.js Update break easyrtc require-js and AMD support #195 

Reviewed and Merged from other: 
- Delete non existing room fix #84
- Updated /etc/init/easyrtc.conf #101
- Added a bower.json #109
- remove duplicate line #117
- Fixed css class for mirroring local video #153	
- fix more trivia: joinRoom()'s roomParameters not a string #168
- fix unused argument in #200
- fix bug for in #201
- fix Invalid range condition in getIthCaller #202
- fix Unnecessary if branch and flow in InitMediaSource #203 
- fix undefined socket.io method #209

Some already merged or closed, but still for reference.
- WebRTC Version Detecting Fails When Simulating iPad In Chrome Dev Tools #90
- callFailureCB is undefined #133
- FileTransfer support on IE10 #137
- FileTransfer support on IE10 #155
 
cc @dryswabbie
reporters @omochi, @ShiYw, @rretter

### CODE REVIEW

File: api/easyrtc.js [hunk 1]
```diff
@@ -3305,39 +3285,43 @@ var Easyrtc = function() {
...
         for (roomName in self.roomData) {
-            mediaIds = self.getRoomApiField(roomName, easyrtcId, "mediaIds");
-            if (!mediaIds) {
-                continue;
-            }
-            for (streamName in mediaIds) {
-                if (mediaIds.hasOwnProperty(streamName) &&
-                        mediaIds[streamName] === webrtcStreamId) {
-                    return streamName;
+            if (self.roomData.hasOwnProperty(roomName)) {
```
Comment (hthetiot) [inline api/easyrtc.js]: Just aded hasOwnProperty, indent cause large diff

File: api/easyrtc_int.js [hunk 2]
```diff
@@ -4666,28 +4720,26 @@ var Easyrtc = function() {
         var added = {}, deleted = {};
         var subPart;
         for (i in newVersion) {
-            if (!newVersion.hasOwnProperty(i)) {
```
Comment (hthetiot) [inline api/easyrtc_int.js]: 15 years of javascript never saw "!newVersion.hasOwnProperty(i)", make my eyes cry.

File: api/easyrtc.js [hunk 3]
```diff
@@ -6310,8 +6318,8 @@ var Easyrtc = function() {
             try {
                self.webSocket = io.connect(serverPath, connectionOptions);
             } catch(socketErr) {
-               errorCallback( self.errCodes.SYSTEM_ERROR,
-                     socketError.toString());
+               errorCallback( self.errCodes.SYSTEM_ERROR, 
+                     socketErr.toString());
                return;
```
Comment (hthetiot) [inline api/easyrtc.js]: So annoying to get the real error, an socketError undefined instead of socketErr into to catch.

File: api/easyrtc.js [hunk 4]
```diff
@@ -6084,7 +6092,7 @@ var Easyrtc = function() {
 
             function iceAddSuccess() {}
             function iceAddFailure(domError) {
-                easyrtc.showError(self.errCodes.ICECANDIDATE_ERR, "bad ice candidate (" + domError.name + "): " +
+                self.showError(self.errCodes.ICECANDIDATE_ERR, "bad ice candidate (" + domError.name + "): " +
```
Comment (hthetiot) [inline api/easyrtc.js]: why do self at the root and not use it. Did that everywhere.

File: api/easyrtc.js [hunk 5]
```diff
@@ -5217,7 +5215,7 @@ var Easyrtc = function() {
             pc = self.createRTCPeerConnection(iceConfig, buildPeerConstraints());
             if (!pc) {
                 message = "Unable to create PeerConnection object, check your ice configuration(" +
-                        JSON.stringify(ice_config) + ")";
+                        JSON.stringify(iceConfig) + ")";
                 if (self.debugPrinter) {
```
Comment (hthetiot) [inline api/easyrtc.js]: Annoying.

File: api/easyrtc.js [hunk 6]
```diff
@@ -4829,7 +4821,7 @@ var Easyrtc = function() {
         // call B as a positive offer to B's offer.
         //
         if (offersPending[otherUser]) {
-            wasAcceptedCB(true);
+            wasAcceptedCB(true, otherUser);
```
Comment (hthetiot) [inline api/easyrtc.js]: Missing params, annoying.

File: api/easyrtc.js [hunk 7]
```diff
@@ -1955,11 +1929,11 @@ var Easyrtc = function() {
      */
     function event(eventName, callingFunction) {
         if (typeof eventName !== 'string') {
-            self.showError(self.errCodes.DEVELOPER_ERR, src + " called without a string as the first argument");
+            self.showError(self.errCodes.DEVELOPER_ERR, callingFunction + " called without a string as the first argument");
```
Comment (hthetiot) [inline api/easyrtc.js]: Again an annoying undefined.

File: .jshintrc [hunk 8]
```diff
@@ -0,0 +1,122 @@
+{
```
Comment (hthetiot) [inline .jshintrc]: And you are lucky! I could have use JSLINT and not JSHINT.

File: api/easyrtc.js [hunk 9]
```diff
@@ -2040,7 +2014,8 @@ var Easyrtc = function() {
         PEER_GONE: "PEER_GONE", // peer doesn't exist
         ALREADY_CONNECTED: "ALREADY_CONNECTED",
         BAD_CREDENTIAL: "BAD_CREDENTIAL",
-        ICECANDIDATE_ERR: "ICECANDIDATE_ERROR"
+        ICECANDIDATE_ERR: "ICECANDIDATE_ERROR",
+        NOVIABLEICE: "NOVIABLEICE_ERROR",
     };
```
Comment (omochi) [inline api/easyrtc.js]: It should be following for consistency with others.

```
ICECANDIDATE_ERR: "ICECANDIDATE_ERR",
NOVIABLEICE: "NOVIABLEICE",
```

**Comment by omochi** [comment]
I think this patch is nice and great work. (except a tiny one I noted)
@EricDavies please accept these commits and clear issues.

**Comment by hthetiot** [comment]
@omochi,  I fixed your review and also applied SIGNAL_ERROR -> SIGNAL_ERR to be the same suffix on every errors "const" - https://github.com/hthetiot/easyrtc/commit/07f550805385c5bb69cccb6450b3464da0ba5830

**Comment by omochi** [comment]
@hthetiot thank you 😀

**Comment by EricDavies** [comment]
Hi Harold,

I've gotten authorization from my boss to give someone from the easyrtc
community administrative access to the easyrtc github site so that they
apply the minor fixes and pull requests that are needed to bring easyrtc
back to a state of usability. Rod and I just snowed under with work on a
critical project  right now and can't do it. Would you be willing to take
on that chore, at least for a few weeks?

Eric.

**Comment by hthetiot** [comment]
Yes @EricDavies I will work with @omochi to merge the issues in beta and prepare the next release for you to merge in master if that works for you.

**Comment by hthetiot** [comment]
Sincerely thanks @EricDavies

**Comment by hthetiot** [comment]
@omochi Ready on my side for one last review.

**Comment by omochi** [comment]
@hthetiot I will check later.

**Comment by hthetiot** [comment]
@omochi Roger that! I used it without issue for now.

**Comment by omochi** [comment]
@hthetiot 

I checked commits following.

> merge Delete non existing room fix #84

Code is ok.

But, I am not sure where this new behavior is good or bad.

When we something delete,
if it is already not exists,
we can think as we get expected result that this is not exists.

I think that consistency is most important thing such case.
So if other deletion apis for entity other than room
returns failure when it is not exists, 
this patch is ok.

> merge Updated /etc/init/easyrtc.conf #101

ok

> merge Added a bower.json #109

ok

> merge remove duplicate line #117

ok

> merge Fixed css class for mirroring local video #153

ok

> fix undefined socket.io method #209

ok. thank you.

> fix Unnecessary if branch and flow in InitMediaSource #203

ok. thank you.

> fix Invalid range condition in getIthCaller #202	

ok. thank you.

> fix bug for in #201

ok. thank you.

> fix unused argument in #200

ok.

But there is regression about `fix bug for in #201`.
please fix it.

> wrong property access of message object for type 'error' in onChannel…
…Cmd #199

ok. thank you.

> fix more trivia: joinRoom()'s roomParameters not a string #168

ok.

> merge New feature: IceConnectionStateChangeListener #122

ok.

> remove SetVideoBandWidth from doc and source after deprecation for mu…
…ltiple release #120

ok.

> error "attempt to connect when already connected to socket server" even if connection failed #98

I wrote my opinion at #98 thread.
Please read it.

> fix Strange code about refreshPane #204	

ok. thank you.

> merge updated example server #112

ok.

**Comment by hthetiot** [comment]
Will check you review ASAP

On Saturday, March 5, 2016, omochimetaru <notifications@github.com> wrote:

> @hthetiot <https://github.com/hthetiot>
>
> I checked commits following.
>
> merge Delete non existing room fix #84
> <https://github.com/priologic/easyrtc/pull/84>
>
> Code is ok.
>
> But, I am not sure where this new behavior is good or bad.
>
> When we something delete,
> if it is already not exists,
> we can think as we get expected result that this is not exists.
>
> I think that consistency is most important thing such case.
> So if other deletion apis for entity other than room
> returns failure when it is not exists,
> this patch is ok.
>
> merge Updated /etc/init/easyrtc.conf #101
> <https://github.com/priologic/easyrtc/pull/101>
>
> ok
>
> merge Added a bower.json #109
> <https://github.com/priologic/easyrtc/pull/109>
>
> ok
>
> merge remove duplicate line #117
> <https://github.com/priologic/easyrtc/pull/117>
>
> ok
>
> merge Fixed css class for mirroring local video #153
> <https://github.com/priologic/easyrtc/pull/153>
>
> ok
>
> fix undefined socket.io method #209
> <https://github.com/priologic/easyrtc/issues/209>
>
> ok. thank you.
>
> fix Unnecessary if branch and flow in InitMediaSource #203
> <https://github.com/priologic/easyrtc/issues/203>
>
> ok. thank you.
>
> fix Invalid range condition in getIthCaller #202
> <https://github.com/priologic/easyrtc/issues/202>
>
> ok. thank you.
>
> fix bug for in #201 <https://github.com/priologic/easyrtc/issues/201>
>
> ok. thank you.
>
> fix unused argument in #200
> <https://github.com/priologic/easyrtc/issues/200>
>
> ok.
>
> But there is regression about fix bug for in #201.
> please fix it.
>
> wrong property access of message object for type 'error' in onChannel…
> …Cmd #199 <https://github.com/priologic/easyrtc/issues/199>
>
> ok. thank you.
>
> fix more trivia: joinRoom()'s roomParameters not a string #168
> <https://github.com/priologic/easyrtc/issues/168>
>
> ok.
>
> merge New feature: IceConnectionStateChangeListener #122
> <https://github.com/priologic/easyrtc/issues/122>
>
> ok.
>
> remove SetVideoBandWidth from doc and source after deprecation for mu…
> …ltiple release #120 <https://github.com/priologic/easyrtc/issues/120>
>
> ok.
>
> error "attempt to connect when already connected to socket server" even if
> connection failed #98 <https://github.com/priologic/easyrtc/issues/98>
>
> I wrote my opinion at #98 <https://github.com/priologic/easyrtc/issues/98>
> thread.
> Please read it.
>
> fix Strange code about refreshPane #204
> <https://github.com/priologic/easyrtc/issues/204>
>
> ok. thank you.
>
> merge updated example server #112
> <https://github.com/priologic/easyrtc/pull/112>
>
> ok.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/priologic/easyrtc/pull/198#issuecomment-192626630>.
>


-- 
Harold Thétiot.

**Comment by hthetiot** [comment]
@omochi 

About #84 
> But, I am not sure where this new behavior is good or bad.
Will double check

About #201 and #200 i fixed regression

I merged the Jslint 190f7d2 that is only large because i changed the order of "private" function and "expression" for not to be use before being declared.

**Comment by hthetiot** [comment]
Ready to merge, will do change log and wait last approval from @omochi and @EricDavies or any other before merging.

**Comment by omochi** [comment]
- About #84
Thank you checking. so ok.

- About #201 and #202 
I checked commit `fix regression on #201 from #200	a28ddbc`. ok.

- About Jslint big pass
Yes I know. ok.

- commit `	fix review #198	e119fed`
I checked it. ok.

--

I think that we can merge it now at last !

**Pull request merged**